### DISCUSSION TITLE
making compatible with other efficient models like falcon, vicuna?

### DISCUSSION THREAD

**Discussion by 77bgh** [imartinez/privateGPT#580] [Ideas]
Hello, how could we make compatible privateGPT,py with these models:

https://huggingface.co/tiiuae/falcon-40b-instruct/tree/main
https://huggingface.co/ausboss/llama-30b-supercot/tree/main
https://huggingface.co/Aeala/VicUnlocked-alpaca-30b/tree/main
https://huggingface.co/TheBloke/Wizard-Vicuna-13B-Uncensored-HF

**Comment by pseudotensor** [comment]
@77bgh We recently added wizardlm (CPU from TheBloke and same GPU from HF support) as well as Falcon support in h2oGPT: https://github.com/h2oai/h2ogpt

**Comment by Zirgite** [comment]
I made it work with a very powerful gpt4-x-vicuna-13B.ggmlv3.q5_1.bin
This is the content of the .env file you need to modify. also do not forget to put gpt4-x-vicuna-13B.ggmlv3.q5_1.bin in the models folder
_______________________________________________________
.env
_______________________________________________________

PERSIST_DIRECTORY=db
MODEL_TYPE=LlamaCpp
MODEL_PATH=models/gpt4-x-vicuna-13B.ggmlv3.q5_1.bin
EMBEDDINGS_MODEL_NAME=all-MiniLM-L6-v2
MODEL_N_CTX=1000
TARGET_SOURCE_CHUNKS=4

I chose this model because it is 4-bit and it is very fast I do not see much speed difference with the original model:
I think you can experiment with the other versions. 
https://huggingface.co/TheBloke/gpt4-x-vicuna-13B-GGML

gpt4-x-vicuna-13B.ggmlv3.q4_1.bin | q4_1 | 4bit | 8.95GB | 10GB | 4-bit. Higher accuracy than q4_0 but not as high as q5_0. However has quicker inference than q5 models.

**Reply by 77bgh** [reply]
pefect, well done, 

Yest I test it , it's working with gpt4-x-vicuna-13B.ggmlv3.q5_1.bin , I see better results than groovy 

:) I see it's working as well with hermes (https://gpt4all.io/index.html) , looks slower than vicuna, but much accurate

PERSIST_DIRECTORY=db
MODEL_TYPE=LlamaCpp
MODEL_PATH=models/ggml-v3-13b-hermes-q5_1.bin
EMBEDDINGS_MODEL_NAME=all-MiniLM-L6-v2
MODEL_N_CTX=1000
TARGET_SOURCE_CHUNKS=4

**Reply by akashlinux10may** [reply]
how to restrict answers only from our source documents?

**Reply by 77bgh** [reply]
this might work ? 


...
# Interactive questions and answers
while True:
    query = input("\nEnter a query: ")
    if query == "exit":
        break

    # Get the answer from the chain
    res = qa(query)
    uploaded_docs = [doc for doc in res['source_documents'] if doc.metadata.get('source') == 'uploaded']
    answer, docs = res['result'], [] if args.hide_source else uploaded_docs

    # Print the result
    print("\n\n> Question:")
    print(query)
    print("\n> Answer:")
    print(answer)

    # Print the relevant sources used for the answer
    for document in docs:
        print("\n> " + document.metadata["source"] + ":")
        print(document.page_content)
...

**Reply by 77bgh** [reply]
an interesting the question would be as well

how can I add learnings from these documents into th emodel (bin file) itself ?

**Reply by akashlinux10may** [reply]
Your suggestion also didnt work. It still answering general queries also like who is Einstein.

**Reply by zirubak** [reply]
I download model from https://huggingface.co/TheBloke/gpt4-x-vicuna-13B-GGML and use the below env. 

PERSIST_DIRECTORY=db
MODEL_TYPE=LlamaCpp
MODEL_PATH=models/gpt4-x-vicuna-13B.ggmlv3.q5_1.bin
EMBEDDINGS_MODEL_NAME=all-MiniLM-L6-v2
MODEL_N_CTX=1000
TARGET_SOURCE_CHUNKS=4

But it has this error. 

llama.cpp: loading model from models/gpt4-x-vicuna-13B.ggmlv3.q5_1.bin
**error loading model: unknown (magic, version) combination: 67676a74, 00000003; is this really a GGML file?**
llama_init_from_file: failed to load model
Traceback (most recent call last):
  File "C:\Users\baek\OneDrive - HP Inc\Documents\TP_Recommendataion\Coding\AI\privateGPT\privateGPT.py", line 82, in <module>
    main()
  File "C:\Users\baek\OneDrive - HP Inc\Documents\TP_Recommendataion\Coding\AI\privateGPT\privateGPT.py", line 36, in main
    llm = LlamaCpp(model_path=model_path, n_ctx=model_n_ctx, n_batch=model_n_batch, callbacks=callbacks, verbose=False)
  File "pydantic\main.py", line 341, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for LlamaCpp
__root__
  Could not load Llama model from path: models/gpt4-x-vicuna-13B.ggmlv3.q5_1.bin. Received error  (type=value_error)

Is it any issue to use gpt4-x-vicuna-13B.ggmlv3.q5_1.bin model?

**Reply by Zirgite** [reply]
try with the other models in the list of available versions. Try first the fastest ones. I am telling you I did not get it in the first try either. 
gpt4-x-vicuna-13B.ggmlv3.q4_0.bin, then  gpt4-x-vicuna-13B.ggmlv3.q4_1.bin

**Reply by Zirgite** [reply]
maybe just prompt the model to use only the reference material and the name of the file.

**Reply by akashlinux10may** [reply]
I tried custom prompt as well but no solution.

**Reply by Zirgite** [reply]
There can be two ways with prompting. The first way is to ask it to use exclusively the available information. The second way is to forbid it to use any other information other than what is contained in the files.  You can try both approaches in the same time. Every model will react differently to this, also if you change the data set it can change also the overall result. At the end you may experiment with different models to find which is best suited for your particular task. I think there are not possibilities to fine tune as in the woogabooga. At the end this tool is extremely powerful and experimental.
Also do not forget the golden rule of prompting. Be specific. If the model does not find the relevant information in the data files it will try to hallucinate it. Include a final prompt to the Model if the information is not available to notify you. The same way try and except command work in Python.

**Comment by 77bgh** [comment]
it's working with different model "paraphrase-MiniLM-L6-v2" , looks faster

however if you ask him :"create in python a df with 2 columns: fist_name and last_name and populate it with 10 fake names, then print the results"

 it will not do it as paraphrase-MiniLM-L6-v2 is different than all-MiniLM-L6-v2 , using all-MiniLM-L6-v2 it will do it, but it's longer the answer

1. "paraphrase-MiniLM-L6-v2":
This model is based on the MiniLM architecture and has 6 layers. It is specifically designed for paraphrasing tasks, which involve generating alternative versions of a given text while preserving the original meaning. This model is fine-tuned on paraphrase data to optimize its performance for such tasks. It is a smaller version of the model compared to the "all-MiniLM" variants.

2. "all-MiniLM-L6-v2":
This model, also based on the MiniLM architecture, consists of 6 layers. It is a general-purpose language model designed to handle various natural language processing tasks. It can be used for tasks such as text classification, sentiment analysis, question answering, and more. However, unlike the "paraphrase-MiniLM-L6-v2" model, it is not specifically optimized for paraphrasing.

3. "all-MiniLM-L12-v2":
This model is similar to the "all-MiniLM-L6-v2" model in terms of being a general-purpose language model based on the MiniLM architecture. However, it is more powerful as it consists of 12 layers, which allows for deeper representations of language and potentially better performance on complex tasks. The additional layers enable the model to capture more intricate patterns and dependencies in the text, resulting in improved language understanding capabilities.


.env 

PERSIST_DIRECTORY=db
MODEL_TYPE=LlamaCpp
MODEL_PATH=models/gpt4-x-alpaca-13b.ggmlv3.q5_1.bin
EMBEDDINGS_MODEL_NAME=paraphrase-MiniLM-L6-v2
MODEL_N_CTX=1400
MODEL_N_BATCH=8
TARGET_SOURCE_CHUNKS=2

**Reply by akashlinux10may** [reply]
are you saying the above model would not answer out the context questions?

**Reply by 77bgh** [reply]
yes and no, needs testing, see above, when I asked it:"create in python a df with 2 columns: fist_name and last_name and populate it with 10 fake names, then print the results", as answer it repeated my q

**Reply by akashlinux10may** [reply]
I could not able to download this model how to use this model and please try on some PDF, confirm that is it answering within context or out of the context.

**Reply by 77bgh** [reply]
I am not using it for testing against uploaded documents for the moment
